agent/cache-types/streaming_health_services.go

```diff
@@ -154,17 +154,18 @@ func (s *healthView) Update(events []*pbsubscribe.Event) error {
 			return fmt.Errorf("unexpected event type for service health view: %T",
 				event.GetPayload())
 		}
-		node := serviceHealth.CheckServiceNode
-		id := fmt.Sprintf("%s/%s", node.Node.Node, node.Service.ID)
 
+		id := serviceHealth.CheckServiceNode.UniqueID()
 		switch serviceHealth.Op {
 		case pbsubscribe.CatalogOp_Register:
-			checkServiceNode := pbservice.CheckServiceNodeToStructs(serviceHealth.CheckServiceNode)
-			s.state[id] = *checkServiceNode
+			csn := pbservice.CheckServiceNodeToStructs(serviceHealth.CheckServiceNode)
+			s.state[id] = *csn
 		case pbsubscribe.CatalogOp_Deregister:
 			delete(s.state, id)
 		}
 	}
+	// TODO(streaming): should this filter be applied to only the new CheckServiceNode
+	// instead of the full map, which should already be filtered.
 	if s.filter != nil {
 		filtered, err := s.filter.Execute(s.state)
```

Comment:
I noticed this earlier, but forgot to call it out. It seems strange that we would run the filter against the entire (potentially very large) map, instead of running it on only the new entry.

Any reason these filters need to be applied to a map? What happens if we filter out the entire `CheckServiceNode`, do we get a nil back ?

Also, it feels like this filtering overlaps some with the `SubscribeRequest.Key`. Right now key filter is done on the server side, and this filtering is client side.

Reply:
I know @banks and I talked about the client vs server side filtering a while back but I forget why there needs to be a difference there.

As for why we need to re-filter the entire map I am not sure. Maybe it was because the filter doesn't expose the underlying `bexpr.Evaluator` which is what you would need to execute against a single `CheckServiceNode` type. I agree though that there doesn't seem much point in re-filtering what has already been filtered previously.